8351369: [macos] Use --install-dir option with DMG packaging

sashamatveev opened the pull request:
- `--install-dir` option in DMG packaging is no longer ignored.
- Defaults are still the same: `/Applications` and `/Library/Java/JavaVirtualMachines`.
- It will be user responsibility to make sure installation directory exist on target machine, since DMG cannot create directories during drag and drop.
- Target directory in case of non-default installation dir will display full path. See image below for example.

![Screenshot 2025-05-28 at 12 07 53 AM](https://github.com/user-attachments/assets/fbcba07b-74a5-4276-90c3-5f8d2d0dcc16)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351369](https://bugs.openjdk.org/browse/JDK-8351369): [macos] Use --install-dir option with DMG packaging (**Enhancement** - P4)


### Reviewers
 * [Alexey Semenyuk](https://openjdk.org/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25481/head:pull/25481` \
`$ git checkout pull/25481`

Update a local copy of the PR: \
`$ git checkout pull/25481` \
`$ git pull https://git.openjdk.org/jdk.git pull/25481/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25481`

View PR using the GUI difftool: \
`$ git pr show -t 25481`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25481.diff">https://git.openjdk.org/jdk/pull/25481.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25481#issuecomment-2915238018)
</details>

alexeysemenyukoracle commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacDmgPackageBuilder.java:
```diff
@@ -54,18 +55,58 @@ List<Path> validatedDmgContent() {
+                             I18N.format("message.install-dir-invalid",
+                             installDir.toString()));
+            }
+
+            if (!Files.exists(installDir)) {
+                try {
+                    // Install dir can have multiple non-existing directories.
+                    // We need to find first non-existing directory, so we can
+                    // delete it.
+                    Path parent = installDir.getParent();
+                    Path deleteRoot = null;
+                    while (parent != null) {
+                        if (!Files.exists(parent.getParent())) {
+                            parent = parent.getParent();
+                        } else {
+                            deleteRoot = parent;
+                            break;
+                        }
+                    }
+
+                    if (parent == null) {
+                        throw new RuntimeException(
+                             I18N.format("message.install-dir-invalid",
+                             installDir.toString()));
+                    }
+
+                    installDirDeleteRoot.path = deleteRoot;
+
+                    try {
+                        Files.createDirectories(installDir);
+                    } catch (Exception ex) {
+                        throw new RuntimeException(
+                             I18N.format("message.install-dir-create",
+                             installDir.toString(),
+                             ex.getLocalizedMessage()));
+                    }
+                } catch (Exception ex) {
+                    Log.verbose(ex);
+                    throw new RuntimeException(ex);
+                }
```
This code should be in a separate function covered with unit tests.

alexeysemenyukoracle replied:
Another issue with this code is that it creates a directory during the configuration phase. Files/directories should be created at the packaging phase. Another reason to move it into MacDmgPackager

alexeysemenyukoracle commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacDmgPackageBuilder.java:
```diff
@@ -54,18 +55,58 @@ List<Path> validatedDmgContent() {
+                             ex.getLocalizedMessage()));
+                    }
+                } catch (Exception ex) {
+                    Log.verbose(ex);
```
What is the point of this log statement? The error will be logged at the top level anyway

alexeysemenyukoracle commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/model/MacDmgPackageMixin.java:
```diff
@@ -32,6 +32,14 @@ public interface MacDmgPackageMixin {
+     *
+     * @return Root of created install dir
+     */
+    Optional<Path> installDirDeleteRoot();
```
Why is this information a part of the model? Can't dmg packager figure it out on its own?

alexeysemenyukoracle commented on test/jdk/tools/jpackage/share/InstallDirTest.java:
```diff
@@ -113,115 +113,56 @@ public static void testLinuxInvalid(String installDir) {
+    public static void testMacCreate(String installDir) {
+        new PackageTest()
+        .setExpectedExitCode(1)
+        .excludeTypes(PackageType.MAC_PKG)
```
`.forTypes(PackageType.MAC_DMG)` would be more appropriate as the test applies to DMG and not to "all packages except of PKG". If we add another bundle type, `.excludedTypes(Package Type.MAC_PKG)` will become wrong.

alexeysemenyukoracle commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/resources/MacResources.properties:
```diff
@@ -80,7 +80,8 @@ message.preparing-scripts=Preparing package scripts.
 message.signing.pkg=Warning: For signing PKG, you might need to set "Always Trust" for your certificate using "Keychain Access" tool.
 message.setfile.dmg=Setting custom icon on DMG file skipped because 'SetFile' utility was not found. Installing Xcode with Command Line Tools should resolve this issue.
-message.install-dir-ignored=Warning: "--install-dir" option is ignored for DMG packaging. The installation directory will default to {0}.
+message.install-dir-invalid=Error: "--install-dir" value {0} is ivalid for DMG packaging. Make sure it is valid path to existing or non-existing directory.
+message.install-dir-create=Error: Unable to create install dir {0}. Make sure you have sufficient permissions or create it manually. Exception: {1}
```
Are these error messages DMG-specific? If "yes", let's give them less vague names, like `message.dmg-install-dir-invalid`

alexeysemenyukoracle commented:
> If the installation directory doesn't exist, jpackage will try to create and delete it right after the DMG package is created.
Is this absolutely necessary? Is there a better way?

alexeysemenyukoracle commented:
It looks like you removed DMG-specific runtime installer tests.

sashamatveev commented:
>> If the installation directory doesn't exist, jpackage will try to create and delete it right after the DMG package is created.

> Is this absolutely necessary? Is there a better way?

Yes, there is better way. I just figure it out. AppleScript alias command cannot create links to non-existing locations, but shell `ln` command can create it. I updated script to use shell to create link. Most code I added is gone. Can you re-review it?

alexeysemenyukoracle commented:
The fix looks neat and clean 👍. Thank you!
